File: src/binds/c/queue/queue.zig
const vk = @import("../vulkan.zig");
const device = @import("../device/device.zig");
const instance = @import("../instance/instance.zig");
const std = @import("std");


pub const FeatureCallback = extern struct {
    callback: *const fn (vk.VkPhysicalDevice, [*c]vk.VkQueueFamilyProperties, c_int, c_int, *anyopaque) callconv(.c) bool,
    args: *anyopaque
};
const queueCreatesInitResult = extern struct {
    create_infos: [*c]vk.VkDeviceQueueCreateInfo,
    priorities: [*c]f32,
    indices: [*c]u32,
    info_map: [*c]u32,
    len: u32,
};
const queueDeviceInitResult = extern struct {
    queues: [*c]vk.VkQueue,
    len: u32,
};
extern fn queueCreatesInit(*const vk.VkPhysicalDevice, [*c]FeatureCallback, c_int) queueCreatesInitResult;
extern fn queueDeviceInit(*const vk.VkDevice, *const queueCreatesInitResult) queueDeviceInitResult;
extern fn queueDeinit(*const queueCreatesInitResult, *const queueDeviceInitResult) void;


pub fn wrap(comptime func: *const fn (vk.VkPhysicalDevice, []vk.VkQueueFamilyProperties, u32, *anyopaque) bool, args: *anyopaque) FeatureCallback {
    var callback: FeatureCallback = undefined;
    callback.args = args;
    callback.callback = struct {
        pub fn inner(physdev: vk.VkPhysicalDevice, props: [*c]vk.VkQueueFamilyProperties, num_props: c_int, idx: c_int, data: *anyopaque) callconv(.c) bool {
            var arr: []vk.VkQueueFamilyProperties = undefined;
            arr.ptr = props;
            arr.len = @intCast(num_props);
            return func(physdev, arr, @intCast(idx), data);
}
    }.inner;
    return callback;
}

pub const queue = struct {
    queues: []vk.VkQueue,
    create_info: []vk.VkDeviceQueueCreateInfo,
    indicies: []u32,
    info_map: []u32, // the nth element of this array tells you which index of queues and create_info that family is
    init_result: queueCreatesInitResult,
    get_result: queueDeviceInitResult,

    pub fn init(dev: *device.device, check_features: []FeatureCallback) !queue {
        var q: queue = undefined;
        try q._pre_dev_init(dev.physical_device.physical_device,  check_features);
        return q;
    }

    fn _pre_dev_init(q: *queue, physdev: vk.VkPhysicalDevice, check_features: []FeatureCallback) !void {
        const result = queueCreatesInit(&physdev, check_features.ptr, @intCast(check_features.len));
        q.create_info.ptr = result.create_infos;
        q.create_info.len = result.len;
        q.indicies.ptr = result.indices;
        q.indicies.len = check_features.len;
        q.info_map.ptr = result.info_map;
        q.info_map.len = check_features.len;
        q.init_result = result;
    }
    
    pub fn get(q: *queue, dev: *device.device) !void {
        const result = queueDeviceInit(&dev.dev, &q.init_result);
        q.queues.ptr = result.queues;
        q.queues.len = result.len;
        q.get_result = result;
    }

    pub fn deinit(q: *queue) void {
        queueDeinit(&q.init_result, &q.get_result);
    }
};

File: src/binds/c/queue/queue.c
#include <stdint.h>
#include <stdlib.h>
#include <vulkan/vulkan.h>
#include <stdbool.h>
#include <assert.h>


struct FeatureCallback {
	bool (*callback)(VkPhysicalDevice, VkQueueFamilyProperties*, int, int, void*);
	void* args;
};

static uint32_t _get_queue_family_index(VkPhysicalDevice physdev, struct FeatureCallback callback) {
	uint32_t num_queue_families = 0;
	vkGetPhysicalDeviceQueueFamilyProperties(physdev, &num_queue_families, NULL);
	VkQueueFamilyProperties queues[num_queue_families];
	vkGetPhysicalDeviceQueueFamilyProperties(physdev, &num_queue_families, queues);
	for (int i = 0; i < num_queue_families; i++) {
		if (callback.callback(physdev, queues, num_queue_families, i, callback.args)) {
			return i;
		}
	}
	assert(false); // assume unreachable
}

typedef struct queueCreatesInitResult {
	VkDeviceQueueCreateInfo* create_infos;
	float* priorities;
	uint32_t* indices;
	uint32_t* info_map;
	uint32_t len;
} queueCreatesInitResult;

queueCreatesInitResult queueCreatesInit(const VkPhysicalDevice* physdev, 
		struct FeatureCallback callbacks[], int num_features) {
	queueCreatesInitResult result;
	result.create_infos = malloc(num_features * sizeof(VkDeviceQueueCreateInfo));
	result.priorities = malloc(num_features * sizeof(float));
	result.indices = malloc(num_features * sizeof(uint32_t));
	result.info_map = malloc(num_features * sizeof(uint32_t));
	uint32_t num_uniques = 0;
	for (int i = 0; i < num_features; i++) {
		uint32_t family_index = _get_queue_family_index(*physdev, callbacks[i]);
		result.indices[i] = family_index;
		result.priorities[i] = 1.0f;
		bool broke = false;
		for (int j = 0; j < i; j++) {
			if (result.create_infos[j].queueFamilyIndex == family_index) {
				broke = true;
				break;
			}
		}
		if (broke) {
			result.info_map[i] = num_uniques - 1;
			continue;
		}
		result.create_infos[num_uniques].sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
		result.create_infos[num_uniques].queueFamilyIndex = family_index;
		result.create_infos[num_uniques].queueCount = 1;
		result.create_infos[num_uniques].pQueuePriorities = &(result.priorities[i]);
		num_uniques++;
		result.info_map[i] = num_uniques - 1;
	}
	result.len = num_uniques;
	return result;
}

typedef struct queueDeviceInitResult {
	VkQueue* queues;
	uint32_t len;
} queueDeviceInitResult;

queueDeviceInitResult queueDeviceInit(const VkDevice* dev, const queueCreatesInitResult* creates) {
	queueDeviceInitResult result;
	result.queues = malloc(creates->len * sizeof(VkQueue));
	for (int i = 0; i < creates->len; i++) {
		vkGetDeviceQueue(*dev, creates->create_infos[i].queueFamilyIndex, i, result.queues);
	}
	result.len = creates->len;
	return result;
}


void queueDeinit(const queueCreatesInitResult* result1, const queueDeviceInitResult* result2) {
	free(result2->queues);
	free(result1->priorities);
	free(result1->create_infos);
	free(result1->indices);
	free(result1->info_map);
}
